kuma/core/jobs.py

```diff
@@ -37,6 +39,46 @@ def hash(self, value):
         return hashlib.md5(six.b(':').join(value)).hexdigest()
 
 
+class GenerationJob(KumaJob):
+    generation_age = 60 * 60 * 24 * 365
+
+    def __init__(self, *args, **kwargs):
+        self.generation_args = kwargs.pop('generation_args', [])
+        super(KumaJob, self).__init__(*args, **kwargs)
+
+    @cached_property
+    def generation_key(self):
+        generation_args = ':'.join([str(key) for key in self.generation_args])
+        return '%s:%s:generation' % (self.class_path, generation_args)
+
+    def key(self, *args, **kwargs):
+        key = super(GenerationJob, self).key(*args, **kwargs)
+        return '%s@%s' % (key, self.generation())
+
+    def generation(self):
```

Comment:
What I mean is that you could use a cacheback job to create and maintain the random generation key, instead of direct cache manipulation.  Something like:

```python
class GenerationJob(KumaJob):
    generation_age = 60 * 60 * 24 * 365

    def __init__(self, *args, **kwargs):
        generation_args = kwargs.pop('generation_args', [])
        super(KumaJob, self).__init__(*args, **kwargs)
        self.generation_key = GenerationKeyJob(
            age=self.generation_age, for_class=self.class_path,
            generation_args=generation_args)

    def key(self, *args, **kwargs):
        key = super(GenerationJob, self).key(*args, **kwargs)
        return '%s@%s' % (key, self.generation_key.get())

    def invalidate_generation(self):
        self.generation_key.delete()


class GenerationKeyJob(Job):
    def __init__(self, age, for_class, generation_args, *args, **kwargs):
        self.age = age
        self.for_class = for_class
        self.generation_args = generation_args

    def key(self, *args, **kwargs):
        generation_args = ':'.join([str(key) for key in self.generation_args])
        return '%s:%s:generation' % (self.for_class, generation_args)

    def fetch(self, *args, **kwargs):
        return crypto.get_random_string(length=12)
```

This isn't cut and paste code, but it only requires understanding cacheback jobs, not Django cache interfaces and other bits.